Dockerfile

```diff
@@ -1,109 +1,146 @@
-    PYTHONDONTWRITEBYTECODE=1 \
-    PYTHONHASHSEED=0
-
-RUN mkdir /workspace
-WORKDIR /workspace
-
-# We are going to use an apt cache on the host, so disable the default debian
+# we are going to use an apt cache on the host, so disable the default debian
 # docker clean up that deletes that cache on every apt install
 RUN rm -f /etc/apt/apt.conf.d/docker-clean
 
+# use deadsnakes ppa to install a fully working base python installation
+# see: https://gist.github.com/tiran/2dec9e03c6f901814f6d1e8dad09528e
+# use space efficient utility from base image
+RUN --mount=type=cache,target=/var/cache/apt \
+    echo "deb https://ppa.launchpadcontent.net/deadsnakes/ppa/ubuntu noble main" > /etc/apt/sources.list.d/deadsnakes-ppa.list && \
+    /usr/lib/apt/apt-helper download-file 'https://keyserver.ubuntu.com/pks/lookup?op=get&search=0xf23c5a6cf475977595c89f51ba6932366a755776' /etc/apt/trusted.gpg.d/deadsnakes.asc
+
 # Add Microsoft package archive for installing MSSQL tooling
-# Add deadsnakes PPA for installing new Python versions
 RUN --mount=type=cache,target=/var/cache/apt \
     echo 'deb [arch=amd64,arm64,armhf signed-by=/usr/share/keyrings/microsoft.asc] https://packages.microsoft.com/ubuntu/24.04/prod noble main' \
       > /etc/apt/sources.list.d/mssql-release.list && \
     /usr/lib/apt/apt-helper download-file \
         "https://packages.microsoft.com/keys/microsoft.asc" \
-        /usr/share/keyrings/microsoft.asc && \
-    echo "deb [signed-by=/usr/share/keyrings/deadsnakes.asc] https://ppa.launchpadcontent.net/deadsnakes/ppa/ubuntu noble main" \
-      > /etc/apt/sources.list.d/deadsnakes-ppa.list && \
-    /usr/lib/apt/apt-helper download-file \
-        'https://keyserver.ubuntu.com/pks/lookup?op=get&search=0xf23c5a6cf475977595c89f51ba6932366a755776' \
-        /usr/share/keyrings/deadsnakes.asc
-
+        /usr/share/keyrings/microsoft.asc
 
-# Install root dependencies, including Python
-COPY dependencies.txt /root/dependencies.txt
-# use space efficient utility from base image
+# install any additional system dependencies
+# NOTE: ensure .python-version is not excluded in .dockerignore
 RUN --mount=type=cache,target=/var/cache/apt \
-    /usr/bin/env ACCEPT_EULA=Y /root/docker-apt-install.sh /root/dependencies.txt
-
-#################################################
+    --mount=type=bind,source=.,target=/app \
+    /usr/bin/env ACCEPT_EULA=Y  /root/docker-apt-install.sh /app/dependencies.txt "python$(cat /app/.python-version)"
+
+# uv env var documentation: https://docs.astral.sh/uv/reference/environment/
+# copy files rather than symlink since the cache and the target are on different filesystems
+ENV UV_LINK_MODE=copy
+# compile at installation time, not import time
+ENV UV_COMPILE_BYTECODE=1
+# we are providing python via deadsnakes ppa
+# (as we require dynamic linking of openssl for security reasons)
+ENV UV_PYTHON_DOWNLOADS=never
+# set the directory for the venv
+ENV UV_PROJECT_ENVIRONMENT="/opt/venv"
```

Comment:
Is there any reason for these not to be set further down, just above where we invoke `uv`?

Reply:
No, only that this is where they are in the repo template. Can move them if that's preferable (and in the repo template too,  if you think it's important?)

Reply:
This is true, but we need them in both build and app images, and setting them twice could lead to inconsistencies, so I think this is less of two weevils